tutorials/tensordictmodule.ipynb

```diff
@@ -0,0 +1,2042 @@
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 36,
+   "id": "ba74542a",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "class TokensToQKV(nn.Module):\n",
+    "    def __init__(self, to_dim, from_dim, latent_dim):\n",
+    "        super().__init__()\n",
+    "        self.q = nn.Linear(to_dim, latent_dim)\n",
+    "        self.k = nn.Linear(from_dim, latent_dim)\n",
+    "        self.v = nn.Linear(from_dim, latent_dim)\n",
+    "    def forward(self, X_to, X_from):\n",
+    "        Q = self.q(X_to)\n",
+    "        K = self.k(X_from)\n",
+    "        V = self.v(X_from)\n",
+    "        return Q, K, V\n",
+    "\n",
+    "class SplitHeads(nn.Module):\n",
+    "    def __init__(self, num_heads):\n",
+    "        super().__init__()\n",
+    "        self.num_heads = num_heads\n",
+    "    def forward(self, Q, K, V):\n",
+    "        batch_size, to_num, latent_dim = Q.shape\n",
+    "        _, from_num, _ = K.shape\n",
+    "        d_tensor = latent_dim // self.num_heads\n",
+    "        Q = Q.reshape(batch_size, to_num, self.num_heads, d_tensor).transpose(1, 2)\n",
+    "        K = K.reshape(batch_size, from_num, self.num_heads, d_tensor).transpose(1, 2)\n",
+    "        V = V.reshape(batch_size, from_num, self.num_heads, d_tensor).transpose(1, 2)\n",
+    "        return Q, K, V\n",
+    "class Attention(nn.Module):\n",
+    "    def __init__(self, latent_dim, to_dim):\n",
+    "        super().__init__()\n",
+    "        self.softmax = nn.Softmax(dim=-1)\n",
+    "        self.out = nn.Linear(latent_dim, to_dim)\n",
+    "    def forward(self, Q, K, V):\n",
+    "        batch_size, n_heads, to_num, d_in = Q.shape\n",
+    "        attn = self.softmax(Q @ K.transpose(2, 3) / d_in)\n",
+    "        out = attn @ V\n",
+    "        out = self.out(out.transpose(1, 2).reshape(batch_size, to_num, n_heads*d_in))\n",
+    "        return out, attn\n",
+    "class SkipLayerNorm(nn.Module):\n",
+    "    def __init__(self, to_len, to_dim):\n",
+    "        super().__init__()\n",
+    "        self.layer_norm = nn.LayerNorm((to_len, to_dim))\n",
+    "    def forward(self, x_0, x_1):\n",
+    "        return self.layer_norm(x_0+x_1)\n",
+    "class FFN(nn.Module):\n",
+    "    def __init__(self, to_dim, hidden_dim, dropout_rate = 0.2):\n",
+    "        super().__init__()\n",
+    "        self.FFN = nn.Sequential(\n",
+    "            nn.Linear(to_dim, hidden_dim),\n",
+    "            nn.ReLU(),\n",
+    "            nn.Linear(hidden_dim, to_dim),\n",
+    "            nn.Dropout(dropout_rate)\n",
+    "        )\n",
+    "    def forward(self, X):\n",
+    "        return self.FFN(X)\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "9651d3c2",
+   "metadata": {},
+   "source": [
+    "Now, we can build the TransformerBlock thanks to the TensorDictModule. Since the changes affect the tensor dict, we just need to map outputs to the right name such as it is picked up by the next block."
```

Comment:
TransformerBlock => `TransformerBlock`

TensorDictModule => `TensorDictModule`

tensor dict => tensordict or `TensorDict`. tensor dict feels like a regular dict of tensors.

Let's be careful with the "we just need": try to avoid saying things are easy, we just need to do this or that

Reply:
Let's drop the "Now" at the beginning of sentences.